tests/fgcmcalTestBase.py

```diff
@@ -551,6 +549,94 @@ def _testFgcmCalibrateTract(self, visits, tract,
+
+        Raises
+        ------
+        Exceptions on test failures.
+        """
+        import numpy.testing as testing
+        import esutil
+
+        # Check the visit catalogs are identical
+        visitCat1 = butler1.get('fgcmVisitCatalog').asAstropy()
+        visitCat2 = butler2.get('fgcmVisitCatalog').asAstropy()
+
+        for col in visitCat1.columns:
+            if isinstance(visitCat1[col][0], str):
+                testing.assert_array_equal(visitCat1[col], visitCat2[col])
+            else:
+                testing.assert_array_almost_equal(visitCat1[col], visitCat2[col])
+
+        # Check that the observation catalogs have the same length
+        starObs1 = butler1.get('fgcmStarObservations')
+        starObs2 = butler2.get('fgcmStarObservations')
+        self.assertEqual(len(starObs1), len(starObs2))
+
+        # Check that the number of stars is the same and all match
+        starIds1 = butler1.get('fgcmStarIds')
+        starIds2 = butler2.get('fgcmStarIds')
+        self.assertEqual(len(starIds1), len(starIds2))
+
+        # htm = esutil.htm.HTM(11)
+        matcher = esutil.htm.Matcher(11, starIds1['ra'], starIds1['dec'])
+        matches = matcher.match(starIds2['ra'], starIds2['dec'], 1./3600., maxmatch=1)
+
+        self.assertEqual(len(matches[0]), len(starIds1))
+
+        # Check that the number of observations of each star is the same
+        testing.assert_array_equal(starIds1['nObs'][matches[1]],
+                                   starIds2['nObs'][matches[0]])
+
+        # And to test the contents, we need to unravel the observations and make
+        # sure that they are matched individually, because the two catalogs
+        # are constructed in a different order
+        starIndices1 = butler1.get('fgcmStarIndices')
+        starIndices2 = butler2.get('fgcmStarIndices')
+
+        test1 = np.zeros(len(starIndices1), dtype=[('ra', 'f8'),
+                                                   ('dec', 'f8'),
+                                                   ('x', 'f8'),
+                                                   ('y', 'f8'),
+                                                   ('psf_candidate', 'b1'),
+                                                   ('visit', 'i4'),
+                                                   ('ccd', 'i4'),
+                                                   ('instMag', 'f4'),
+                                                   ('instMagErr', 'f4'),
+                                                   ('jacobian', 'f4')])
+        test2 = np.zeros_like(test1)
+
+        ctr = 0
+        obsIndex = starIndices1['obsIndex']
+        for i in range(len(starIds1)):
+            ind = starIds1['obsArrIndex'][matches[1][i]]
+            nObs = starIds1['nObs'][matches[1][i]]
+            for name in test1.dtype.names:
+                test1[name][ctr: ctr + nObs] = starObs1[name][obsIndex][ind: ind + nObs]
+            ctr += nObs
```

Comment:
I think these two loops could be turned into one method or closure, so the shared logic would be more obvious. How exactly the `numpy_util.match` would work isn't obvious to me, but the behavior of that bit is also not entirely obvious to me either.